cmd/kubeadm/app/util/marshal.go

```diff
@@ -66,67 +69,74 @@ func UnmarshalFromYamlForCodecs(buffer []byte, gv schema.GroupVersion, codecs se
 	return runtime.Decode(decoder, buffer)
 }
 
-// ExtractAPIVersionAndKindFromYAML extracts the APIVersion and Kind fields from YAML bytes
-func ExtractAPIVersionAndKindFromYAML(b []byte) (string, string, error) {
-	decoded, err := LoadYAML(b)
-	if err != nil {
-		return "", "", fmt.Errorf("unable to decode config from bytes: %v", err)
-	}
-
-	kindStr, ok := decoded["kind"].(string)
-	if !ok || len(kindStr) == 0 {
-		return "", "", fmt.Errorf("any config file must have the kind field set")
-	}
-	apiVersionStr, ok := decoded["apiVersion"].(string)
-	if !ok || len(apiVersionStr) == 0 {
-		return "", "", fmt.Errorf("any config file must have the apiVersion field set")
+// SplitYAMLDocuments reads the YAML bytes per-document, unmarshals the TypeMeta information from each document
+// and returns a map between the GroupVersionKind of the document and the document bytes
+func SplitYAMLDocuments(yamlBytes []byte) (map[schema.GroupVersionKind][]byte, error) {
+	gvkmap := map[schema.GroupVersionKind][]byte{}
+	buf := bytes.NewBuffer(yamlBytes)
+	reader := utilyaml.NewYAMLReader(bufio.NewReader(buf))
+	for {
+		typeMetaInfo := runtime.TypeMeta{}
+		// Read one YAML document at a time, until io.EOF is returned
+		b, err := reader.Read()
+		if err == io.EOF {
+			break
+		} else if err != nil {
+			return nil, err
+		}
+		if len(b) == 0 {
+			break
+		}
+		// Deserialize the TypeMeta information of this byte slice
+		if err := yaml.Unmarshal(b, &typeMetaInfo); err != nil {
+			return nil, err
+		}
+		// In earlier kubeadm versions TypeMeta wasn't set correctly when marshalling, so empty TypeMeta means the v1alpha1
```

Comment:
We wont support untyped object, fail fast

Reply:
fixed